### PR TITLE
Unit Test for DeckNameComparator

### PR DESCRIPTION
**Pull request by AyaanJaved**
Added Unit Testing for DeckNameComparator class.

## Purpose / Description
To test the DeckNameComparator class

## Approach
Added 3 methods - 

1. To check comparision lexicographically
2. To check comparision by length
3. To check comparision when strings are equal

## How Has This Been Tested?

Performed Unit test on my local machine

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Before
+    public void setUp(){
+        deckNameComparator = new DeckNameComparator();
+    }
+
+    //Tests the comparision of two strings Lexicographically (Dictionary Order)
+    @Test
+    public void compareLexicographically() {
+       assertThat(deckNameComparator.compare("Aa", "Az"), is(-25));
```
[the remark this answers is not in the document]
Comment (AyaanJaved) [reply AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:20]: My bad, we should only be checking if its less than zero...gonna rectify it

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.ichi2.utils;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class DeckNameComparatorTest {
```
[the remark this answers is not in the document]
Comment (AyaanJaved) [reply AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:11]: > * How case sensitivity compares using other languages - again: Anki Desktop is the source of truth here
Noted, working on it

**Comment by Arthur-Milchior** [comment]
In my opinion, having a single test function in DectTest would be better. It takes a lot of spaces and are so simple and indépendant that it’s not worth creating so many comparators object. 

Better idea would be to take an array, in increasing order, of list of deck name. Each element of the list should be equal. Then iterate on each pair, checking that all results are as expected

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java [hunk 3]
```diff
@@ -14,21 +16,29 @@ public void setUp() {
...
+        assertThat(deckNameComparator.compare("ab", "BB"), is(lessThan(0)));
+    }
+
+    @Test
+    public void compareCaseGreaterThan() {
+        assertThat(deckNameComparator.compare("AB", "Aa"), is(greaterThan(0)));
+    }
+
+    //Test to further check implementation of case insensitivity.
+    @Test
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:40]: `@Ignore("#8397")` if it currently fails

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java [hunk 4]
```diff
@@ -14,21 +16,29 @@ public void setUp() {
         deckNameComparator = new DeckNameComparator();
     }
 
-    //Tests the comparision of two strings Lexicographically (Dictionary Order)
     @Test
-    public void compareLexicographically() {
-       assertThat(deckNameComparator.compare("Aa", "Az"), is(-25));
+    public void compareLessThan() {
+       assertThat(deckNameComparator.compare("aa", "aa:bb"), is(lessThan(0)));
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:21]:
```suggestion
       assertThat(deckNameComparator.compare("aa", "aa:bb"), lessThan(0));
```

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java (hunk 3 above)
Comment (AyaanJaved) [reply AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:40]: Definitely seems strange to me...I couldn't find which part of code is causing Issue #8397 becuase the comparator passes all the case tests.

**Comment by AyaanJaved** [comment]
> Better idea would be to take an array, in increasing order, of list of deck name. Each element of the list should be equal. Then iterate on each pair, checking that all results are as expected

Sure thing

**Comment by Arthur-Milchior** [comment]
> Sure thing

Let us know when it's done then.

**Comment by AyaanJaved** [comment]
I implemented the unit test function by sorting array of DeckNames using the DeckNameComparator.

Having difficulty in writing tests for checking case sensitivity in other languages.
Should I create new arrays following the same pattern for each language?
`{"AA", "ab", "BB"}`

**Comment by Arthur-Milchior** [comment]
You ca ’t use sort in test. The sorting function assume tge order is well defined, and that is actually what you want to test. 
Why is the test with a subdeck in a special case ?


Can you look upstream what they do internationaly ? We want to copy them. If they only us US rules, that will be what we’ll do

**Comment by AyaanJaved** [comment]
> You ca ’t use sort in test. The sorting function assume tge order is well defined, and that is actually what you want to test.

I might be wrong, but the sorting function will only give correct results if the comparator is implemented correctly?
> Why is the test with a subdeck in a special case ?

My bad, will rectify it in the next commit
> Can you look upstream what they do internationaly ? We want to copy them. If they only us US rules, that will be what we’ll do

I checked for other languages upstream, they follow the same pattern as the US one.

File: AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java [hunk 5]
```diff
@@ -16,21 +16,12 @@ public void setUp() {
         deckNameComparator = new DeckNameComparator();
     }
 
-    //Testing DeckNameComparator by sorting an array of different string size.
+    //Testing DeckNameComparator by sorting an array of deck names.
     @Test
-    public void sortDecksBySize() {
-        String[] deckNames = new String[]{"aa", "aa::bb"};
+    public void sortDeckNames() {
+        String[] deckNames = new String[]{"AA", "ab", "BB", "aa::bb", "aa::ab"};
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/utils/DeckNameComparatorTest.java:22]: also test `aa` without the subdeck

Also check for 3 levels - caught a bug introduced in a PR recently where the comparator was only set to handle two levels

**Comment by Arthur-Milchior** [comment]
> > You ca ’t use sort in test. The sorting function assume tge order is well defined, and that is actually what you want to test.
> 
> I might be wrong, but the sorting function will only give correct results if the comparator is implemented correctly?

No.
Le’s say you have 3 valued a b and c, with a <b<c<a
Then the function may compare a to b, then b to c, and chhose the order a b c
It may also compare a to c, then a to b and choose the order c a b

If your order is not really and order,  sorting may return undertministic list

**Review by david-allison-1 — APPROVED** [review]
Looks good to me

To note: `aa` and `AA` should be the same deck

**Comment by Arthur-Milchior** [comment]
> Looks good to me
> 
> To note: `aa` and `AA` should be the same deck

That's why I mentioned  "array of list of deck name". Using equivalence class in each list. So that the logic allows to consider the fact that any order is acceptable between aa and AA

**Comment by kuroni** [comment]
Sorry for the intrusion, I want to add some of my thoughts on this issue.

Instead of sorting, one approach is to have a pre-sorted list, then we loop through every ordered pair and check their comparison.

For example, say you test using the sorted list `{"AA", "aa", "aa::ab"}`. Then we will check the comparison between these pairs: `{"AA", "AA"}`, `{"AA", "aa"}`, `{"AA", "aa:ab"}`, `{"aa", "AA"}`, `{"aa", "aa"}`, `{"aa", "aa:ab"}`, `{"aa:ab", "AA"}`, `{"aa:ab", "aa"}`, `{"aa:ab", "aa:ab"}`.

There might be some handling at the equality cases, but in my opinion this is one way to approach the test if you don't want to sort.

**Comment by Arthur-Milchior** [comment]
That is almost what I tried to suggest indeed. Except that I would have done:

`{{"A"}, {"AA", "aa"}, {"aa::ab"}}`

**Review by mikehardy — APPROVED** [review]
I think this is fine - I'm not sure where the comments were directed but sending a list of possible deck names into the comparator and then verifying they are in the order we know they should be (which is what this test does now) is how to do the test

Could the array contents be different? Of course, fixture data can be infinitely variable - if there are some special cases missing (foreign languages? case sensitivity tests?) then they can be added as a follow-on, sure

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for April 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in April)

Thanks!

**Pull request merged**